pkg/streamingpromql/types/data.go

```diff
@@ -27,6 +28,47 @@ type InstantVectorSeriesData struct {
 	Histograms []promql.HPoint
 }
 
+type InstantVectorSeriesDataIterator struct {
+	data   *InstantVectorSeriesData
+	fIndex int
+	hIndex int
+}
+
+func NewInstantVectorSeriesDataIterator(data *InstantVectorSeriesData) *InstantVectorSeriesDataIterator {
+	return &InstantVectorSeriesDataIterator{
+		data: data,
+	}
+}
+
+// Next returns either a float or histogram iterating through both sets of points.
+// It returns the next point with the lowest timestamp.
+// If h is not nil, the value is a histogram, otherwise it is a float.
+// If no more values exist ok is false.
+func (i *InstantVectorSeriesDataIterator) Next() (t int64, f float64, h *histogram.FloatHistogram, ok bool) {
+	if i.fIndex >= len(i.data.Floats) && i.hIndex >= len(i.data.Histograms) {
+		return 0, 0, nil, false
+	}
+
+	if i.fIndex >= len(i.data.Floats) {
+		i.hIndex++
+		return i.data.Histograms[i.hIndex-1].T, 0, i.data.Histograms[i.hIndex-1].H, true
```

Comment:
[nit] The logic here might be slightly clearer with something like this:

```suggestion
		point := i.data.Histograms[i.hIndex]
		i.hIndex++
		return point.T, 0, point.H, true
```

(same applies below)

Reply:
I did this to avoid having to assign point, but the -1 still does an op so for readability I'll change it back

Reply:
I suspect the compiler will turn both into the same thing.